staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -883,7 +910,19 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 	}
 	span.AddEvent("Filtered items", attribute.Int("count", listVal.Len()))
 	if c.versioner != nil {
-		if err := c.versioner.UpdateList(listObj, readResourceVersion, "", nil); err != nil {
+		var remainingItemCount *int64
+		var continueValue string
+		if applyLimit && len(objs) > len(selectedObjects) {
```

Comment:
Could you help me understand this? Can we not check `len(selectedObjects) > opts.Predicate.Limit`?

Also, `objs` comes from `c.ListItems`, which gets it from:
https://github.com/kubernetes/kubernetes/blob/b0263baae51895058b64b8e6f510a22a14ed8d3d/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L502

Here we try and get the first index that matches as a best effort scenario:
https://github.com/kubernetes/kubernetes/blob/b0263baae51895058b64b8e6f510a22a14ed8d3d/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L519-L528

and then do final filtering here:
https://github.com/kubernetes/kubernetes/blob/b0263baae51895058b64b8e6f510a22a14ed8d3d/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L869

So it might just be the case that `len(objs) == len(selectedObjects)`.

Reply:
I think we should be fine since `WaitUntilFreshAndList` doesn't apply any filters, as the comment says:

```
// This isn't the place where we do "final filtering" - only some "prefiltering" is happening here. So the only
// requirement here is to NOT miss anything that should be returned. We can return as many non-matching items as we
// want - they will be filtered out later. The fact that we return less things is only further performance improvement.
```